python_ta/checkers/one_iteration_checker.py

```diff
@@ -25,20 +26,63 @@ class OneIterationChecker(BaseChecker):
     # pass in message symbol as a parameter of check_messages
     @check_messages('one-iteration')
     def visit_for(self, node):
-        if self._check_one_iteration(node):
+        if hasattr(node, 'cfg_block') and self._check_one_iteration(node):
             self.add_message('one-iteration', node=node)
 
     @check_messages('one-iteration')
     def visit_while(self, node):
-        if self._check_one_iteration(node):
+        if hasattr(node, 'cfg_block') and self._check_one_iteration(node):
             self.add_message('one-iteration', node=node)
 
     def _check_one_iteration(self, node: Union[astroid.For, astroid.While]) -> bool:
-        """Return whether the given loop is guaranteed to stop after one iteration."""
-        # TODO: Complete this method.
-        # print(node.cfg_block)
+        """Return whether the given loop is guaranteed to stop after one iteration.
 
-        return False
+        More precisely, Returns False if there exists a direct predecessor
+        block `p` to the start of the loop block `s` such that the
+        first statement in `p` is a child node of <node> and that there exists a
+        path from `s` to `p.
+
+        Note: For `while` loops, 'start of the loop block' refers to the block with
+        the test condition (or the first of the blocks that make up test condition).
+        For `for` loops, it refers to the block with the assignment target.
+        """
+        start = node.target if isinstance(node, astroid.For) else node
+        preds = start.cfg_block.predecessors
+
+        if preds == []:
+            return False
+
+        for pred in preds:
+            pred = pred.source
+            if node.parent_of(pred.statements[0]) and \
+                    self.is_predecessor(pred, start.cfg_block):
+                return False
+        return True
+
+    def is_predecessor(self, block: CFGBlock, _pred: CFGBlock) -> bool:
+        """Returns True if <_pred> is a predecessor of <block>.
+
+        A block <d> is predecessor of block <e> if there exists a path from <d>
+        to <e>.
+
+        Preconditions:
+            - <_pred> is the start block of the loop.
+            - <block>.statements[0] is a child node of the astroid.For/While node.
+
+        Assumption:
+            - Since every path from the root to <block> must pass through
+        <_pred>, we only need to check one of the paths to see if
+        <_pred> is the predecessor of <block>.
+
+        """
+        if block is _pred:
+            return True
+        elif block.predecessors == []:
+            return False
+        else:
+            # The left-most predecessor ensures we won't step into an
+            # unreachable block of a nested loop.
```

Comment:
This holds because the start block of a while/for subgraph has to already have a predecessor: https://github.com/pyta-uoft/pyta/blob/9980e44f051722abc45d0afead2ca87539c8f27d/python_ta/cfg/visitor.py#L116

This implies that the edge linking to an unreachable block will not be the first element in the `predecessors` list.

Reply:
Correction: This does not hold for all cases. For instance when having statements after jump statements within a nested `If` statement. I've made the necessary changes.